common/src/main/java/io/crate/types/TypeSignatures.java

```diff
@@ -0,0 +1,214 @@
+        }
+
+        return TypeCompatibility.incompatible();
+    }
+
+    @Nullable
+    private static DataType<?> convertTypeByPrecedence(DataType<?> arg1, DataType<?> arg2) {
+        final DataType<?> higherPrecedenceArg;
+        final DataType<?> lowerPrecedenceArg;
+        if (arg1.precedes(arg2)) {
+            higherPrecedenceArg = arg1;
+            lowerPrecedenceArg = arg2;
+        } else {
+            higherPrecedenceArg = arg2;
+            lowerPrecedenceArg = arg1;
+        }
+
+        final boolean lowerPrecedenceCastable = lowerPrecedenceArg.isConvertableTo(higherPrecedenceArg);
+        final boolean higherPrecedenceCastable = higherPrecedenceArg.isConvertableTo(lowerPrecedenceArg);
+
+        if (lowerPrecedenceCastable) {
+            return higherPrecedenceArg;
+        } else if (higherPrecedenceCastable) {
+            return lowerPrecedenceArg;
+        }
+
+        return null;
+    }
+
+    @Nullable
+    private static DataType<?> coerceTypeBase(DataType<?> sourceType, String resultTypeBase) {
+        DataType<?> resultType = getType(parseTypeSignature(resultTypeBase));
+        if (resultType.equals(sourceType)) {
+            return sourceType;
+        }
+        return convertTypeByPrecedence(sourceType, resultType);
+    }
+
+    private static boolean isCovariantParametrizedType(DataType<?> type) {
+        // if we ever introduce contravariant, this function should be changed to return an enumeration: INVARIANT, COVARIANT, CONTRAVARIANT
+        return type instanceof ObjectType || type instanceof ArrayType;
+    }
+
+    private static TypeCompatibility typeCompatibilityForCovariantParametrizedType(DataType<?> fromType, DataType<?> toType) {
+        ArrayList<TypeSignatureParameter> commonParameterTypes = new ArrayList<>();
+        List<DataType<?>> fromTypeParameters = fromType.getTypeParameters();
+        List<DataType<?>> toTypeParameters = toType.getTypeParameters();
+
+        if (fromTypeParameters.size() != toTypeParameters.size()) {
+            return TypeCompatibility.incompatible();
+        }
+
+        boolean coercible = true;
+        for (int i = 0; i < fromTypeParameters.size(); i++) {
+            TypeCompatibility compatibility = compatibility(fromTypeParameters.get(i), toTypeParameters.get(i));
+            if (!compatibility.isCompatible()) {
+                return TypeCompatibility.incompatible();
+            }
+            coercible &= compatibility.isCoercible();
+            commonParameterTypes.add(TypeSignatureParameter.of(compatibility.getCommonSuperType().getTypeSignature()));
+        }
+        String typeBase = fromType.getTypeSignature().getBase();
+        return TypeCompatibility.compatible(
+            getType(new TypeSignature(typeBase, Collections.unmodifiableList(commonParameterTypes))),
+            coercible);
+    }
+
+    public static class TypeCompatibility {
```

Comment:
Maybe make this top-level. I think if we move the `getType` into `TypeSignature` we could get rid of the `TypeSignatures` "utils" style class and expose the `getCommonSuperType` as part of the (then top-level) `TypeCompatibility` class. This would make the responsibility imho a bit clearer.